File: 백준/Gold/20040. 사이클 게임/사이클 게임.py
import sys

n,m=map(int, sys.stdin.readline().split())
parent=[i for i in range(n+1)]

cnt=0
num=0

def find(x):
    if parent[x]==x:
        return x
    parent[x]=find(parent[x])
    return parent[x]

def union(a,b):
    pa=find(a)
    pb=find(b)

    if pa==pb:
        return
    else:
        if pa<pb:
            parent[pb]=pa
        else:
            parent[pa]=pb
    return

for _ in range(m):
    a,b=map(int, sys.stdin.readline().split())

    cnt+=1
    if find(a)==find(b):
        num=cnt
        break
    union(a,b)

print(num)

File: 백준/Gold/20040. 사이클 게임/사이클 게임.java
import java.io.*;
import java.util.*;



public class Main {

    static int[] parent;
    static int find(int x){
        if (parent[x]==x)
            return x;
        parent[x]=find(parent[x]);
        return parent[x];
    }

    static void union(int a, int b){
        int pa=find(a);
        int pb=find(b);

        if (pa==pb)
            return;
        if (pa<pb)
            parent[pb]=pa;
        else parent[pa]=pb;
        return;
    }



    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st=new StringTokenizer(br.readLine());
        int n=Integer.parseInt(st.nextToken());
        int m=Integer.parseInt(st.nextToken());
        parent=new int[n+1];
        for(int i=0;i<n+1;i++){
            parent[i]=i;
        }
        int num=0;
        for(int i=0;i<m;i++){
            st=new StringTokenizer(br.readLine());
            int a=Integer.parseInt(st.nextToken());
            int b=Integer.parseInt(st.nextToken());

            if (find(a)==find(b)){
                num=i+1;
                break;
            }
            union(a,b);

        }
        System.out.println(num);


    }

}